handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/available/AvailableProductMovesEndpoint.scala

```diff
@@ -44,44 +53,121 @@ object AvailableProductMovesEndpoint {
-      _ <- ZIO.log("Sub: " + sub.toString)
-    } yield Success(List(
-      MoveToProduct(
-        id = "123",
-        name = "Digital Pack",
-        billing = Billing(
-          amount = Some(1199),
-          percentage = None,
-          currency = Currency.GBP,
-          frequency = Some(TimePeriod(TimeUnit.month, 1)),
-          startDate = Some("2022-09-21")
-        ),
-        trial = Some(Trial(14)),
-        introOffer = Some(Offer(
-          Billing(
-            amount = None,
-            percentage = Some(50),
-            currency = Currency.GBP,//FIXME doesn't make sense for a percentage
-            frequency = None,//FIXME doesn't make sense for a percentage
-            startDate = Some("2022-09-21")
-          ),
-          duration = TimePeriod(TimeUnit.month, 3)
-        ))
-      )
-    ))
+
+      // Kick off catalogue fetch in parallel
+      zuoraProductCatalogueFetch <- GetCatalogue.get.fork
+      subscription <- GetSubscription.get(subscriptionName)
+
+      // Next payment date
+      chargedThroughDate <- ZIO.fromOption(subscription.ratePlans.head.ratePlanCharges.head.chargedThroughDate).orElseFail(s"chargedThroughDate is null for subscription $subscriptionName.")
+
+      account <- GetAccount.get(subscription.accountNumber)
+      paymentMethod <- GetAccount.getPaymentMethod(account.basicInfo.defaultPaymentMethod.id)
+
+      today <- Clock.currentDateTime.map(_.toLocalDate)
+
+      /*
+      Only show the switch if:
+        credit card is not expired (according to zuora)
+        User is not in payment failure or has unpaid invoices
+        Currency is GBP (initially on day 1 only?)
+        Monthly contribution
+        Only have one sub in the account
+      */
+      isEligible <-
+        (for {
+          _ <- eligibilityCheck(account.subscriptions.length > 1, s"More than one subscription for account for subscription: $subscriptionName")
+          _ <- eligibilityCheck(subscription.ratePlans.head.productRatePlanId != monthlyContributionRatePlanId,s"Subscription: $subscriptionName is not a monthly contribution")
+          _ <- eligibilityCheck(subscription.ratePlans.head.ratePlanCharges.head.currency != "GBP", s"Subscription: $subscriptionName not in GBP")
+          _ <- eligibilityCheck(paymentMethod.NumConsecutiveFailures > 0, s"User is in payment failure with subscription: $subscriptionName")
+          _ <- eligibilityCheck(account.basicInfo.defaultPaymentMethod.creditCardExpirationMonth.isBefore(today), s"card expired for subscription: $subscriptionName")
+        } yield ()).isSuccess
+
+      availableProductMoves <- if (isEligible) {
+        for {
+          zuoraProductCatalogue <- zuoraProductCatalogueFetch.join
+
+          productRatePlans = getAvailableSwitchRatePlans(zuoraProductCatalogue, List("Digital Pack"))
+          moveToProduct <- ZIO.collectAll { productRatePlans.map(x => MoveToProduct.buildResponseFromRatePlan(subscriptionName, x, chargedThroughDate)) }
```

Comment:
this line is a bit hard to read, not sure if we can split it up a bit and use foreach (the type signature looks right for here https://javadoc.io/static/dev.zio/zio_2.12/2.0.0/zio/ZIO$.html#foreach[R,E,A,B,Collection[+Element]%3C:Iterable[Element]](in:Collection[A])(f:A=%3Ezio.ZIO[R,E,B])(implicitbf:zio.BuildFrom[Collection[A],B,Collection[B]],implicittrace:zio.Trace):zio.ZIO[R,E,Collection[B]] )
```suggestion
          moveToProduct <- ZIO.foreach(productRatePlans) { productRatePlan =>
            MoveToProduct.buildResponseFromRatePlan(subscriptionName, productRatePlan, chargedThroughDate)
          }
```

Reply:
yep, it works

Reply:
nice.  so collectAll = sequence and forech = traverse, depending on whether you speak ZIO or cats style.